pkg/networkservice/mechanisms/memif/server.go

```diff
@@ -55,36 +53,28 @@ func NewServer(vppConn api.Connection, options ...Option) networkservice.Network
 func (m *memifServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
 	postponeCtxFunc := postpone.ContextWithValues(ctx)
 
+	// if direct memif we need pass mechanism to client further in request chain
+	if mechanism := memifMech.ToMechanism(request.GetConnection().GetMechanism()); mechanism != nil && m.directMemifEnabled {
+		storeDirectMemifInfo(ctx, directMemifInfo{})
+	}
+
 	conn, err := next.Server(ctx).Request(ctx, request)
 	if err != nil {
 		return nil, err
 	}
 
-	if mechanism := memifMech.ToMechanism(conn.GetMechanism()); mechanism != nil {
-		// Direct memif if applicable
-		if memifSocketAddDel, ok := load(ctx, true); ok && m.directMemifEnabled {
-			_, ok := ifindex.Load(ctx, true)
-			if ok {
-				if err := del(ctx, conn, m.vppConn, true); err != nil {
-					if closeErr := m.closeOnFailure(postponeCtxFunc, conn); closeErr != nil {
-						err = errors.Wrapf(err, "connection closed with error: %s", closeErr.Error())
-					}
-					return nil, err
-				}
-				mechanism.SetSocketFileURL((&url.URL{Scheme: memifMech.SocketFileScheme, Path: memifSocketAddDel.SocketFilename}).String())
-				delete(ctx, true)
-				ifindex.Delete(ctx, true)
-				return conn, nil
-			}
-		}
-	}
-
-	if err := create(ctx, conn, m.vppConn, metadata.IsClient(m)); err != nil {
+	// if direct memif case - just set memif socket name in connection.Mechanism
+	// if not direct memif case - create memif as always
+	dirMemifInfo, ok := loadDirectMemifInfo(ctx)
+	if mechanism := memifMech.ToMechanism(conn.GetMechanism()); mechanism != nil && m.directMemifEnabled && ok && len(dirMemifInfo.socketURL) > 0 {
```

Comment:
Looks like `ok` already means that we have `m.directMemifEnabled`.
```suggestion
	if mechanism := memifMech.ToMechanism(conn.GetMechanism()); mechanism != nil && ok && len(dirMemifInfo.socketURL) > 0 {
```

Reply:
ok, done